### PR TITLE
tools: add cancel-delete store command for pd-ctl

### LINKED ISSUE #4098: support abort deleting a store
### Describe the feature you'd like

It's possible that user accidentally deletes a undesired store. Currently, there is an undocumented way to abort the operation by:
```
curl -X POST http://<ip>:<port>/pd/api/v1/store/<store_id>/state?state=Up
```
. The operation is dangerous as it allows changing state manually. A more safe way is providing a public and well documented command/API to abort the operation.

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Close #4098

### What is changed and how it works?

```commit-message
```

```
» store --help
manipulate or query stores

Usage:
  pd-ctl store [command] [flags]
  pd-ctl store [command]

Available Commands:
  cancel-delete    cancel delete the store
  check            Check all the stores with specified status
  delete           delete the store
  label            set a store's label value
  limit            show or set a store's rate limit
  limit-scene      show or set the limit value for a scene
  remove-tombstone remove tombstone record if only safe
  weight           set a store's leader and region balance weight

Flags:
  -h, --help            help for store
      --jq string       jq query
      --state strings   state filter

Use "pd-ctl store [command] --help" for more information about a command.
```

First, delete store 2:
```
» store delete 2
Success!
```
```
"store": {
   "id": 2,
   "state_name": "Offline"
}
```
Then, cancel delete store 2:
```
» store cancel-delete 2
Success!
```

```
"store": {
   "id": 2,
   "state_name": "Up"
}
```

### Check List

Tests

- Unit test

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):

### Release note

```release-note
None
```

### CODE REVIEW

File: tests/pdctl/store/store_test.go [hunk 1]
```diff
@@ -276,6 +277,17 @@ func (s *storeTestSuite) TestStore(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(strings.Contains(string(output), "Unknown state: Invalid_state"), IsTrue)
 
+	// store undelete <store_id> command
+	args = []string{"-u", pdAddr, "store", "undelete", "1"}
```
Comment (lhy1024) [inline tests/pdctl/store/store_test.go:281]: revote-delete? revote-offline?

Comment (rleungx) [reply tests/pdctl/store/store_test.go:281]: Do you mean revoke or revert?

Comment (lhy1024) [reply tests/pdctl/store/store_test.go:281]: I think `revoke` is better than `undelete`

**Comment by HunDunDM** [comment]
I think it is necessary to confirm whether the Store state is `Offline` to prevent users from using pd-ctl to connect to the old version PD and change the Store state from `Tombstone` to `Up`.

**Comment by rleungx** [comment]
> I think it is necessary to confirm whether the Store state is `Offline` to prevent users from using pd-ctl to connect to the old version PD and change the Store state from `Tombstone` to `Up`.

make sense

**Comment by BusyJay** [comment]
Perhaps "cancel" is a more precise word for this situation.

**Comment by rleungx** [comment]
> Perhaps "cancel" is a more precise word for this situation.

I think both of them are ok to me.

**Comment by BusyJay** [comment]
Revoke usually means take away, like revoke privileges or licenses. Cancel means abort, stop doing sth.

File: tools/pd-ctl/pdctl/command/store_command.go [hunk 2]
```diff
@@ -304,54 +327,124 @@ func deleteStoreCommandFunc(cmd *cobra.Command, args []string) {
...
+	}
+
+	prefix := fmt.Sprintf(storePrefix, args[0])
+	r, err := doRequest(cmd, prefix, http.MethodGet)
+	if err != nil {
+		cmd.Printf("Failed to get store: %s\n", err)
+		return
+	}
+
+	storeInfo := struct {
```
Comment (HunDunDM) [inline tools/pd-ctl/pdctl/command/store_command.go:361]: Can the defined struct be used here?

File: tools/pd-ctl/pdctl/command/store_command.go [hunk 3]
```diff
@@ -304,54 +327,124 @@ func deleteStoreCommandFunc(cmd *cobra.Command, args []string) {
...
+			Address string `json:"address"`
+			State   int32  `json:"state"`
+		} `json:"store"`
+	}{}
+	if err = json.Unmarshal([]byte(r), &storeInfo); err != nil {
+		cmd.Printf("Failed to parse store info: %s\n", err)
+		return
+	}
+
+	if storeInfo.Store.State != 1 {
```
Comment (HunDunDM) [inline tools/pd-ctl/pdctl/command/store_command.go:373]: `1` It's so magical.

File: tools/pd-ctl/pdctl/command/store_command.go (hunk 2 above)
Comment (rleungx) [reply tools/pd-ctl/pdctl/command/store_command.go:361]: We can but I don't think it's necessary.

**Review by HunDunDM — APPROVED** [review]

**Comment by BusyJay** [comment]
This is not enough, store limit should be reset to normal after canceling.

**Comment by HunDunDM** [comment]
Since it was not a reversible operation before, the value before offline will not be recorded. Will it be a problem if it always changes back to 15?

**Comment by rleungx** [comment]
> Since it was not a reversible operation before, the value before offline will not be recorded. Will it be a problem if it always changes back to 15?

I think it needs further discussion, hold for now.

**Review by nolouch — APPROVED** [review]

**Review by nolouch — CHANGES_REQUESTED** [review]
Let us change back to the default store limit first?

**Comment by rleungx** [comment]
> Let us change back to the default store limit first?

I don't think it's a good option if the users change the all stores' limit before.

**Comment by rleungx** [comment]
After discussing with our PM, he suggests adding a limit option when delete or cancel-delete store. 
e.g.
```
store delete 1 1000(represent the limit)
store cancel-delete 1 15(represent the limit)
```
WDYT @BusyJay @nolouch

**Comment by nolouch** [comment]
can we save the previous store limit?

**Review by lhy1024 — COMMENTED** [review]
Will we continue to advance it?

**Comment by BusyJay** [comment]
I don't have strong opinion on how to restore previous settings.

**Review by nolouch — APPROVED** [review]
lgtm

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1277,23 +1278,26 @@ func (c *RaftCluster) UpStore(storeID uint64) error {
...
-	if limiter[storelimit.AddPeer] != nil {
-		options = append(options, core.ResetStoreLimit(storelimit.AddPeer, limiter[storelimit.AddPeer].Rate()))
-	}
-	if limiter[storelimit.RemovePeer] != nil {
-		options = append(options, core.ResetStoreLimit(storelimit.RemovePeer, limiter[storelimit.RemovePeer].Rate()))
+	// get the previous store limit recorded in memory
+	limiter, exist := c.prevStoreLimit[storeID]
+	if exist {
+		options = append(options,
+			core.ResetStoreLimit(storelimit.AddPeer, limiter[storelimit.AddPeer]),
```
Comment (JmPotato) [inline server/cluster/cluster.go:1286]: Don't we need to check whether the `limiter[storelimit.AddPeer]` is nil anymore?

Comment (rleungx) [reply server/cluster/cluster.go:1286]: If the store limit has existed, add peer won't be nil.

**Review by JmPotato — APPROVED** [review]

**Pull request merged**